# **01: Exploratory Data Analysis**

In [73]:
import numpy as np
import pandas as pd

In [74]:
df = pd.read_csv("../data/online_retail_II.csv")

In [75]:
df.shape

(1067371, 8)

In [76]:
df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


## 1.1 Look for null or missing values

In [77]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1067371 entries, 0 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype  
---  ------       --------------    -----  
 0   Invoice      1067371 non-null  str    
 1   StockCode    1067371 non-null  str    
 2   Description  1062989 non-null  str    
 3   Quantity     1067371 non-null  int64  
 4   InvoiceDate  1067371 non-null  str    
 5   Price        1067371 non-null  float64
 6   Customer ID  824364 non-null   float64
 7   Country      1067371 non-null  str    
dtypes: float64(2), int64(1), str(5)
memory usage: 65.1 MB


In [78]:
df.isnull().sum()

Invoice             0
StockCode           0
Description      4382
Quantity            0
InvoiceDate         0
Price               0
Customer ID    243007
Country             0
dtype: int64

In [79]:
df[df["Description"].isnull()].head()


,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
470,489521,21646,NaN,-50,2009-12-01 11:44:00,0.0,NaN,United Kingdom
3114,489655,20683,NaN,-44,2009-12-01 17:26:00,0.0,NaN,United Kingdom
3161,489659,21350,NaN,230,2009-12-01 17:39:00,0.0,NaN,United Kingdom
3731,489781,84292,NaN,17,2009-12-02 11:45:00,0.0,NaN,United Kingdom
4296,489806,18010,NaN,-770,2009-12-02 12:42:00,0.0,NaN,United Kingdom


In [80]:
df[df["Customer ID"].isnull()].head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
263,489464,21733,85123a mixed,-96,2009-12-01 10:52:00,0.00,NaN,United Kingdom
283,489463,71477,short,-240,2009-12-01 10:52:00,0.00,NaN,United Kingdom
284,489467,85123A,21733 mixed,-192,2009-12-01 10:53:00,0.00,NaN,United Kingdom
470,489521,21646,NaN,-50,2009-12-01 11:44:00,0.00,NaN,United Kingdom
577,489525,85226C,BLUE PULL BACK RACING CAR,1,2009-12-01 11:49:00,0.55,NaN,United Kingdom


In [81]:
null_desc = len(df[df["Description"].isnull()])
null_id = len(df[df["Customer ID"].isnull()])
zero_price = len(df[df["Price"]==0])
neg_price = len(df[df["Price"]<0])
print(f"{null_desc} rows with null Description;\n{null_id} rows with null Customer ID;\n{zero_price} rows with Price=0.00;\n{neg_price} rows with negative Price.")

4382 rows with null Description;
243007 rows with null Customer ID;
6202 rows with Price=0.00;
5 rows with negative Price.


In [82]:
# Inspect correlation between zero-price and missing values
all_conditions = len(df[(df["Customer ID"].isnull()) & (df["Description"].isnull()) & (df["Price"] == 0)])
with_description = len(df[(df["Price"]==0) & (df["Customer ID"].isnull())]) - all_conditions
just_zero_price = len(df[(df["Price"]==0) & (df["Description"].notnull()) & (df["Customer ID"].notnull())])

In [83]:
print(f"Of ZERO-PRICED rows:\n{all_conditions} have both Description and Customer ID missng;")
print(f"{with_description} are missing just the Customer ID;")
print(f"{just_zero_price} have both Description and Customer ID.")

Of ZERO-PRICED rows:
4382 have both Description and Customer ID missng;
1749 are missing just the Customer ID;
71 have both Description and Customer ID.


In [84]:
# Zero-priced and no Customer ID
df[(df["Price"]==0) & (df["Customer ID"].isnull()) & (df["Description"].notnull())].head(10)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
263,489464,21733,85123a mixed,-96,2009-12-01 10:52:00,0.0,NaN,United Kingdom
283,489463,71477,short,-240,2009-12-01 10:52:00,0.0,NaN,United Kingdom
284,489467,85123A,21733 mixed,-192,2009-12-01 10:53:00,0.0,NaN,United Kingdom
3162,489660,35956,lost,-1043,2009-12-01 17:43:00,0.0,NaN,United Kingdom
3168,489663,35605A,damages,-117,2009-12-01 18:02:00,0.0,NaN,United Kingdom
4538,489820,21133,invcd as 84879?,-720,2009-12-02 13:23:00,0.0,NaN,United Kingdom
5904,489861,DOT,DOTCOM POSTAGE,1,2009-12-02 14:50:00,0.0,NaN,United Kingdom
6556,489899,79323GR,sold as gold,-954,2009-12-03 09:41:00,0.0,NaN,United Kingdom
6911,490007,84347,21494,-720,2009-12-03 12:09:00,0.0,NaN,United Kingdom
9308,490130,21493,lost?,-600,2009-12-03 18:28:00,0.0,NaN,United Kingdom


In [85]:
# Just zero-priced
df[(df["Price"]==0) & (df["Description"].notnull()) & (df["Customer ID"].notnull())].head(10)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
4674,489825,22076,6 RIBBONS EMPIRE,12,2009-12-02 13:34:00,0.0,16126.0,United Kingdom
6781,489998,48185,DOOR MAT FAIRY CAKE,2,2009-12-03 11:19:00,0.0,15658.0,United Kingdom
16107,490727,M,Manual,1,2009-12-07 16:38:00,0.0,17231.0,United Kingdom
18738,490961,22065,CHRISTMAS PUDDING TRINKET POT,1,2009-12-08 15:25:00,0.0,14108.0,United Kingdom
18739,490961,22142,CHRISTMAS CRAFT WHITE FAIRY,12,2009-12-08 15:25:00,0.0,14108.0,United Kingdom
32916,492079,85042,ANTIQUE LILY FAIRY LIGHTS,8,2009-12-15 13:49:00,0.0,15070.0,United Kingdom
40101,492760,21143,ANTIQUE GLASS HEART DECORATION,12,2009-12-18 14:22:00,0.0,18071.0,United Kingdom
47126,493761,79320,FLAMINGO LIGHTS,24,2010-01-06 14:54:00,0.0,14258.0,United Kingdom
48342,493899,22355,"CHARLOTTE BAG , SUKI DESIGN",10,2010-01-08 10:43:00,0.0,12417.0,Belgium
57619,494607,21533,RETRO SPOT LARGE MILK JUG,12,2010-01-15 12:43:00,0.0,16858.0,United Kingdom


In [86]:
# Just Customer ID Missing
df[(df["Customer ID"].isnull()) & (df["Description"].notnull()) & (df["Price"]>0)].head(10)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
577,489525,85226C,BLUE PULL BACK RACING CAR,1,2009-12-01 11:49:00,0.55,NaN,United Kingdom
578,489525,85227,SET/6 3D KIT CARDS FOR KIDS,1,2009-12-01 11:49:00,0.85,NaN,United Kingdom
1055,489548,22271,FELTCRAFT DOLL ROSIE,1,2009-12-01 12:32:00,2.95,NaN,United Kingdom
1056,489548,22254,FELT TOADSTOOL LARGE,12,2009-12-01 12:32:00,1.25,NaN,United Kingdom
1057,489548,22273,FELTCRAFT DOLL MOLLY,3,2009-12-01 12:32:00,2.95,NaN,United Kingdom
1058,489548,22195,LARGE HEART MEASURING SPOONS,1,2009-12-01 12:32:00,1.65,NaN,United Kingdom
1059,489548,22131,FOOD CONTAINER SET 3 LOVE HEART,2,2009-12-01 12:32:00,1.95,NaN,United Kingdom
1060,489548,22079,RIBBON REEL HEARTS DESIGN,10,2009-12-01 12:32:00,1.65,NaN,United Kingdom
1061,489548,22138,BAKING SET 9 PIECE RETROSPOT,3,2009-12-01 12:32:00,4.95,NaN,United Kingdom
1062,489548,22147,FELTCRAFT BUTTERFLY HEARTS,2,2009-12-01 12:32:00,1.45,NaN,United Kingdom


In [87]:
# Negative price
df[df["Price"]<0]

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
179403,A506401,B,Adjust bad debt,1,2010-04-29 13:36:00,-53594.36,NaN,United Kingdom
276274,A516228,B,Adjust bad debt,1,2010-07-19 11:24:00,-44031.79,NaN,United Kingdom
403472,A528059,B,Adjust bad debt,1,2010-10-20 12:04:00,-38925.87,NaN,United Kingdom
825444,A563186,B,Adjust bad debt,1,2011-08-12 14:51:00,-11062.06,NaN,United Kingdom
825445,A563187,B,Adjust bad debt,1,2011-08-12 14:52:00,-11062.06,NaN,United Kingdom


Upon first inspection we notice a few things:        

1. All rows with missing *Description*, also have null *Customer ID* and zero *Price*: these $4382$ rows may be removed, since too much info is missing for a relevant 
analysis.        

2. $1749$ rows with missing *Customer ID* also have zero price. Upon inspection, it appears that most of these rows refer to damaged or lost products, and even those who have valid Description are already missing too much info (Customer ID and price) so I chose to remove them altogether from further analysis.    

3. The rows in which ONLY *Customer ID* is null are $236876$; they may be kept since they retain meaningful information for statistical analysis on sales. They can be kept out of the analysis if later on we decide to focus on a customer-based analysis.    

4. The rows in which all info is present but the *Price* is set to $0$ are $71$. Of these rows, some have descriptions that point to noise like "Manual" or "test". These may be removed.    

5. $5$ rows have negative price, and they all have the same description: "Adjust bad debt". These can be safely removed as they correspond to accounting adjustments and do not represent actual transactions.

## 1.2 Remove noise

#### **Removing negative-price & missing Description rows**

In [88]:
prev_len = len(df) # for checking

# Remove negative price rows
df = df[df["Price"]>=0]

# Remove rows with missing Description (also missing Customer ID and having zero-price)
df = df[df["Description"].notnull()]

print(f"Number of data samples expected: {prev_len-null_desc-neg_price}\nNumber of data samples: {len(df)}")

Number of data samples expected: 1062984
Number of data samples: 1062984


#### **Removing zero-priced + missing Customer ID rows**

In [90]:
prev_len = len(df) # for checking

# Removing rows with zero price and missing Customer ID
mask = (df["Price"]==0) & (df["Customer ID"].isnull())
df = df[~mask]

print(f"Number of data samples expected: {prev_len-with_description}\nNumber of data samples: {(len(df))}")

Number of data samples expected: 1061235
Number of data samples: 1061235


#### **Removing test or manual samples**
This step focuses on the remainder of the data samples. Inspecting rows that only have zero price, we noticed that some rows were to be considered invalid based on their Description (eg. "test¨, "manual",...) or the Stock Code that was differently formatted.     
This is valid not only for zero-priced rows but the whole dataset, so in this section we focus on removing invalid samples from the whole remaining dataset.    
    

Removing rows with specific Descripions may be too hard to implement, since "Manual" and "Test" are just some examples of invalid Descriptions and we cannot afford to look at all of them.    
Most valid samples have all uppergase Description, but I preferred to not use this criterium since it may be too strict (removed all rows where a unit of measurement like grams or cm is present in the description).    
    

The Stock Code may be more indicative.


In [99]:
# are there non alphanumeric stock codes?
mask = (df["StockCode"].str.isalnum())
df[~mask].head(10)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
18410,C490943,BANK CHARGES,Bank Charges,-1,2009-12-08 14:08:00,15.00,16703.0,United Kingdom
18466,490948,BANK CHARGES,Bank Charges,1,2009-12-08 14:29:00,15.00,16805.0,United Kingdom
31079,491969,gift_0001_80,Dotcomgiftshop Gift Voucher £80.00,1,2009-12-14 17:57:00,69.56,NaN,United Kingdom
32048,491971,gift_0001_20,Dotcomgiftshop Gift Voucher £20.00,2,2009-12-14 18:37:00,17.39,NaN,United Kingdom
33435,C492206,BANK CHARGES,Bank Charges,-1,2009-12-15 16:32:00,848.43,NaN,United Kingdom
40904,492782,gift_0001_10,Dotcomgiftshop Gift Voucher £10.00,1,2009-12-18 17:06:00,8.69,NaN,United Kingdom
40905,492782,gift_0001_20,Dotcomgiftshop Gift Voucher £20.00,1,2009-12-18 17:06:00,17.39,NaN,United Kingdom
41263,492783,gift_0001_50,Dotcomgiftshop Gift Voucher £50.00,1,2009-12-18 17:15:00,43.48,NaN,United Kingdom
45212,493404,gift_0001_30,Dotcomgiftshop Gift Voucher £30.00,1,2009-12-23 15:30:00,26.08,NaN,United Kingdom
45213,493404,gift_0001_20,Dotcomgiftshop Gift Voucher £20.00,1,2009-12-23 15:30:00,17.39,NaN,United Kingdom


In [101]:
# How are lengths distributed?
df["StockCode"].str.len().value_counts().sort_index()

StockCode
1       1708
2        279
3       1439
4       2150
5     928107
6     125819
7       1385
8        100
9         68
12       180
Name: count, dtype: int64

In [103]:
df[df["StockCode"].str.len().isin([1,2])].head(10)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
735,C489535,D,Discount,-1,2009-12-01 12:11:00,9.00,15299.0,United Kingdom
736,C489535,D,Discount,-1,2009-12-01 12:11:00,19.00,15299.0,United Kingdom
2697,489609,M,Manual,1,2009-12-01 14:50:00,4.00,NaN,United Kingdom
3053,C489651,M,Manual,-1,2009-12-01 16:48:00,5.10,17804.0,United Kingdom
5897,C489859,M,Manual,-1,2009-12-02 14:45:00,69.57,NaN,United Kingdom
9259,C490126,M,Manual,-1,2009-12-03 18:12:00,5.95,15884.0,United Kingdom
9292,490127,C2,CARRIAGE,1,2009-12-03 18:13:00,50.00,14156.0,EIRE
9307,C490129,M,Manual,-1,2009-12-03 18:26:00,1998.49,15482.0,United Kingdom
11310,490300,M,Manual,1,2009-12-04 14:19:00,0.85,12970.0,United Kingdom
11311,490300,M,Manual,1,2009-12-04 14:19:00,0.21,12970.0,United Kingdom


We gather two informations:  
1. It's highly probable that all valid codes are alphanumeric, since non-alphanumeric codes correspond to invalid rows;
2. The great majority of samples has a stock code of length 5-6

In [95]:
# Check how many there are in the 71-sized subset
len(df[(df["Price"]==0) & (df["Description"].notnull()) & (df["Customer ID"].notnull()) & (df["StockCode"].str.isalnum())])

71